agent/app/errors.go

```diff
@@ -24,6 +24,18 @@ func isTransient(err error) bool {
 	return ok
 }
 
+type terminalError struct {
+	error
+}
+
+// isTerminal returns true if the error is already wrapped as an unrecoverable condition
+// which will allow agent to exit terminally.
+func isTerminal(err error) bool {
+	// Check if the error is already wrapped as a terminalError
+	_, terminal := err.(terminalError)
```

Comment:
Could there be a situation in which this gets wrapped again? Why not use a deeper equality check like [errors.Is](https://pkg.go.dev/errors#Is) which inspects the error tree rather than its first layer?

Reply:
There should not be a case where it gets wrapped again - the terminal wrap should be the last layer as it's handled in this file. So in this case, both errors.Is and the current behaves the same.

Since existing isTransient is implemented similarly - hopefully this optimization shouldn't be a blocker?

Reply:
This is not a blocker, but in the case where we re-use these errors it may be a better long term change to wrap/unwrap these with the errors library.